src/main/java/io/axoniq/axonserver/connector/query/impl/QueryChannelImpl.java

```diff
@@ -451,10 +467,12 @@ private void doHandleQuery(QueryProviderInbound query, ReplyChannel<QueryRespons
     }
 
     private void doHandleQuery(QueryRequest query, ReplyChannel<QueryResponse> responseChannel) {
-        CompletableFuture<?> completionHandle = new CompletableFuture<>();
-        queriesInProgress.add(completionHandle);
-        completionHandle.whenComplete((r, e) -> queriesInProgress.remove(completionHandle));
-        ReplyChannel<QueryResponse> responseHandler = new CloseAwareReplyChannelAdapter(responseChannel, () -> completionHandle.complete(null));
+        MultiFlowControl multiFlowControl = new MultiFlowControl();
+        Runnable removeQuery = () -> queriesInProgress.remove(query.getMessageIdentifier());
+        QueryInProgress queryInProgress = new QueryInProgress(removeQuery, multiFlowControl);
+        queriesInProgress.putIfAbsent(query.getMessageIdentifier(), queryInProgress);
```

Comment:
The use of `putIfAbsent`  suggests that the given identifier may already exist. In that case, wouldn't we want to use the already existing `QueryInProgess` instance? Of maybe reject the request altogether?

Reply:
Semantically, I think we want to reject the request. If there is already **the same query** in progress, responses sent are valid, and responses to be sent are going to be valid. If you want a new result stream, issue a new query. If, however, receiving **the same query** twice is a result of a retry, rejecting it is also fine.